### ISSUE TITLE
AGS4/New compiler: Feature proposal: Dynamic arrays of dynamic arrays ("jagged" arrays)

### ISSUE THREAD

**Issue by fernewelten** [adventuregamestudio/ags#1258]
So what about making it possible for dynamic array elements to be dynamic arrays?

To declare them, you'd write:
```
int A[][] = new int[7][];
```

In the Engine, this would create a dynamic array of managed elements, each of which would be null (yet).
![pic1](https://user-images.githubusercontent.com/41213694/114701712-722e8680-9d23-11eb-8d2f-bec9beadbeb1.png)

So you'd continue with

```
for (int idx = 0; idx < 7; idx++)
    A[idx] = new int[idx + 1];
```

In the Engine, this would create seven dynamic arrays of non-managed elements, here specifically integers. It's entirely possible that they have different sizes, as in this example, but of course, you can also have arrays of equal size.
![pic3](https://user-images.githubusercontent.com/41213694/114705997-b3756500-9d28-11eb-9901-efcea2195315.png)

So if you go, `player.Say("Get me %d", A[3][2]);`, this would say the value of the integer marked in red.
For context, this is what C# speak calls a "jagged array".
This can be continued to any number of `[]`

Okay, I think I can get the compiler to do it. 

But will the Engine play along? As far as I can see, n-e-a-r-l-y.

- The engine already knows about dynamic arrays of managed elements, and as far as I can see, it will handle the reference counting correctly. 
- Within the arrays, it will use object handles (small integers) instead of object addresses, and serializing/deserializing won't change these handles. So serializing/deserializing won't break the referencing.

However, 
- When a reference count of a dynamic array drops to zero and the array is recycled, then this needs to be propagated repeatedly to the entities it references. Just _one_ step up won't be enough. I _think_ it does so by recursion, but I haven't made quite sure. This can be changed, but even if it doesn't, this mightn't be a complete deal breaker if the garbage collection will see to it that all the reference-less objects are recycled __eventually__.
- When the Engine serializes the managed entities, it does so by ascending handle numbers. That may not be good enough. It may need to serialize the entities that a dynamic array references _before_ serializing the array itself. 

I think that both issues are resolvable with moderate effort and with little risk of introducing hard-to-find bugs.

We'd get a sort of "poor man's" way to define multidimensional arrays out of it. No need to code cumbrous index functions and circumlocutions in your AGS code such as `postbox[postbox_idx(row, column)]`; you could have `postbox[row][column]` directly.

**Comment by rofl0r** [comment]
i can understand that you're excited about your compiler and eager to add new features, though did anybody request these features ? if not this could become something known as feature creep. it's likely only 1 out of 100 users will use such feature once available, but then you have to support it for all eternity.
also imo it's usually preferable to implement stuff like this as library code.

e.g.
```c
struct List *list = ListCreate();
for (x = 0; x < numitems ; x += 1) list.append(ListCreate());
```
voila, you have your 2d array. to access it `list.get(x).get(y)` or something like that.
i'm pretty sure AGS already has/supports similar constructs.

**Comment by ivan-mogilko** [comment]
The biggest problem with AGS script at the moment is lack of type description. Which leads to managed object system *not knowing which elements of a struct are reference counted pointers*. Which in turn means that if you have a pointer to second object inside a managed struct - these will never be properly released - leading to literal memory leaks as engine won't know that it must dispose a no longer referenced managed object.

That is why unfortunately it's either impossible or difficult to make something like the library pseudocode posted above. There may be ugly workarounds, but if we speak of a normal way of doing things - that list inside list currently cannot be done within ags script.

I think dynamic array are the only thing that actually tells that it contains pointers (and it may contain only pointers or only not pointers - so it's simplier). Because of that it seems feasible to do dynamic arrays within dynamic arrays. All the concerns mentioned by @fernewelten are valid though so a research is due.

Speaking of uses, well, only a small portion of users are AGS are scripting something advanced, but many of them would appreciate this possibility. We had multidimensional dyn array requests for years, and every advanced scripter learning to AGS eventually asks about this.

That said, *personally* I'd vote to research solutions for managed pointers inside managed structs first - because that would cover much more use cases (even let users to create their own multidimensional containers in script).
In any case it's important to keep this generic case in mind when solving any recursive disposal / serialization issue, and not create a special solution that works only for dynamic arrays.

**Comment by fernewelten** [comment]
> That said, personally I'd vote to research solutions for managed pointers inside managed structs first 

Well, I've spent some thoughts about this, too. It's still  __very__ preliminary and tentative. But here's what I have been able to find out so far.

From the vantage point of the engine commands: 
- Programs already sort of "_declare_" whether a memory location contains a handle: That's because you can only use a location for a handle if you store the handle via `SCMD_MEMWRITEPTR`. So a `SCMD_MEMWRITEPTR` command is tantamount to saying "Here, right _here_ is going to be a handle from now on".
- Currently, programs even "release" specific memory cells from containing a pointer, by clearing the cell via `SCMD_MEMZEROPTR`. And if they don't, we can mandate that. Thus a `SCMD_MEMZEROPTR` command would be tantamount to saying "From now on, this location _here_ will no longer contain a handle"
- So if the engine wanted to keep track of just __where__ the active handles are, it could. It is already being told.
- An alternative would be  to change the `SCMD_NEWUSEROBJECT` command  in such a way that it tells where the handles are in the object. The compiler would first allocate and initialize a memory chunk with an indication where the handles are, Let's call this chunk a "type structure". The compiler would perhaps use the normal opcodes for allocating and initializing a "type structure". Then it would issue a `SCMD_NEWUSEROBJECT` command, referencing the "type structure" in one of the parameters or registers.

From the vantage point of the engine structures:
- If we need to keep track of where the pointers are, this is going to cost memory in one way or another. After all, to keep track of this information, it must be stored _somewhere_. 
- As @ivan-mogilko pointed out, currently we only have structures that _only_ contain pointers and structures that do _not_ keep track of pointers at all. The latter are the problem.
- If we go the way of evaluating the `SCMD_MEMWRITEPTR` and `SCMD_MEMZEROPTR` commands:
    - we are going to learn piecemeal where the handles currently reside. That makes it kind of hard to keep the information with the memory objects.
    - On the other hand, we _only_ need to store this information with the respective memory objects when we serialize. During program execution, we are free to store the information some other way that is efficient to compute and to use. 
- If we let the compiler declare a "type structure" of the memory objects, 
    - then many objects may share the same "type structure". 
    - For each object, we could store a handle to the "type structure", e.g., at `object[-4]` and then treat that reference in exactly the same way that we treat other dynamic memory: Reference-count it and recycle it as soon as it isn't referenced any longer.

From the vantage point of the data model:
- Once that we introduce memory objects that contain pointers, more complex data structures become possible. In particular, you can have an object A that contains a pointer to object B, and object B that contains a pointer to object A. We'd have to ponder the consequences of that. 

As I said above, these thoughts are very tentative. I'm sure they can be improved on, and in any case, they'd need to be elaborated.

**Comment by ivan-mogilko** [comment]
Well, I would not want to start the discussion about that right here, this belongs to a separate ticket. But only to quickly note -

> Programs already sort of "declare" whether a memory location contains a handle
> So if the engine wanted to keep track of just where the active handles are, it could. It is already being told.

This refers only to regular structs allocated in global memory or stack. In these cases pointers (handles) are managed by compiler itself. But this is never done for managed structs obviously and therefore not relevant to the issue.

> An alternative would be to change the SCMD_NEWUSEROBJECT command in such a way that it tells where the handles are in the object

This was the variant that I was suggesting to people years ago, and according to my thoughts at the time it would be enough for the engine to handle things itself, just like it handles disposal of dynamic array. In short, idea was to write runtime types information as an additional segment in scripts. Engine would keep the table of these types, and NEWUSEROBJECT would accept the ID of the type as an argument.
I can write a ticket with my thoughts after I recall them later.

**Comment by ivan-mogilko** [comment]
I opened a ticket on RTTI: #1259

**Comment by jeancallisti** [comment]
Get rid of pointers. They're an artefact from the past, only useful if you want to be super efficient. They're perfect for C++, but they're not suitable for a high-level scripting language. Use references everywhere. AGS should aim at C#/Java/Python comfort, not C/C++.

**Comment by fernewelten** [comment]
> Side note : Get rid of pointers.

AFAICS, AGS doesn't have raw pointers in the `C` sense: the language lacks any means of taking the address of an arbitrary entity. For instance, there just isn't any way to do `int *a, b; a = &b;` in AGS. 

The diagrams above show pointers because they take the vantage point of the bytecode interpreter. __After__ compilation, even references of high-level languages are generally implemented through a pointer mechanism.

By the way, even the Engine (the bytecode interpreter of AGS) has a memory handling system that is specially secured.

**Comment by jeancallisti** [comment]
That's my point. They're not pointers so let's stop calling them pointers and let's stop using the star notation (which has been removed from AGS4 as someone explained to me on the forums). The word "pointer" is completely foreign to anyone who started programming after 2005.

**Comment by ivan-mogilko** [comment]
Implemented by #2106.

Example of use:
```
short Test[][]; 
int game_start()
{
    // Initialise a 3 by 3 array of arrays
    Test = new short[3][];
    for (int i = 0; i < 3; i++)
        Test[i] = new short[3];
    // Use an element
    Test[2][1] = 4711;
}
```
